### PR TITLE
aliasing fixes

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66977

Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

### CODE REVIEW

**Review by dzhulgakov — COMMENTED** [review]
Update native/README.md with what this syntax means?

**Review by albanD — COMMENTED** [review]
Why do you need a new syntax here?
The previous syntax has the same meaning no?

**Review by eellison — COMMENTED** [review]
I updated and extended the views logic, but we should check in about refactoring the logic. some of the assumptions, like the fact that an alias annotation is present means there's a view, doesn't hold up with the aliasing grammar. 

For instance: `Tensor(a) -> Tensor(b)` should be fine but trigger a view here. I think what we need to do is factor out the reader of schemas in alias_analysis to be seprately queryable.

**Review by soulitzer — COMMENTED** [review]
Thanks for updating the autograd fallback logic to work with the new aliasing grammar. The changes look good, I just had some small comments.

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp [hunk 1]
```diff
@@ -41,6 +41,10 @@ void _foreach_tensor(
   }
 }
 
+bool isTensorOfLists(const at::TypePtr& type) {
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:44]: There is a `isTensorList` method on ivalues of the corresponding other method you are talking about

**Review by albanD — COMMENTED** [review]
I didn't check the not implemented fallback in details as @soulitzer knows that better.
Is there any perf implication of these modifications though?

File: aten/src/ATen/native/README.md [hunk 2]
```diff
@@ -250,6 +250,13 @@ There is also another situation in which we use annotations, namely views.
   - `transpose(Tensor(a) self, int dim0, int dim1) -> Tensor(a)`
     An alias to the memory represented by `self` may be also returned, however it is not mutated.
 
+When a Tensor views are contained in a Tensor list, we need to represent that the output list
+contains Tensors that alias the input.
+  - `func: chunk(Tensor(a -> *) self, int chunks, int dim=0) -> Tensor[]`
+We assume lists contain memory which aliases the heap, so in order to correctly set up the aliasing
```
[the remark this answers is not in the document]
Comment (eellison) [reply aten/src/ATen/native/README.md:256]: I don't want to re-duplicate the alias docs everywhere... will add link/ref

File: tools/codegen/model.py [hunk 3]
```diff
@@ -883,19 +885,30 @@ class Annotation:
...
 
     @staticmethod
     def parse(ann: str) -> 'Annotation':
-        m = re.match(r'^([a-z])(!?)(!?)$', ann)
+        # Only handling afterSet == Wildcard for now
+        becomes_wildcard_index = ann.find(" -> *")
+        if becomes_wildcard_index != -1:
+            after_set = "*"
+            # TODO: im not good enough with regexes to ignore -> *
+            m = re.match(r'^([a-z])(!?)(!?)$', ann[:becomes_wildcard_index] + ann[becomes_wildcard_index + len(" -> *"):])
```
[the remark this answers is not in the document]
Comment (eellison) [reply tools/codegen/model.py:897]: foreach functions dont have this problem because the input lists are assumed to be in the heap

**Review by eellison — COMMENTED** [review]
@albanD haven't tested perf yet but the long term plan would be to cache per operator (see https://github.com/issues/). Inevitably as we add more of the aliasing schema parsing to this pathway (see https://github.com/pytorch/pytorch/issues/30526) perf will slightly degrade, but IMO for that we should add a cache not over-optimize the parsing.

**Comment by bdhirsh** [comment]
There's an `is_view_op` method defined in `model.py` here: https://github.com/pytorch/pytorch/pull/66977/files#diff-a34ee043aa7bde7b608aa03bdd3ff6d24af2fea53f96c01be304034c30a8bfcdR487 that I think you'll need to change- right now it determines aliasing ops based on whether the output tensor has alias annotations.

By the way, what's the reason that we don't need/want alias annotations on the return argument anymore for those multi-tensor-output views? (e.g. split)

**Comment by eellison** [comment]
@bdhirsh the internal aliasing syntax that was used wasn't even being parsed, (e.g. Tensor(a)[] == Tensor()[]). it was just the fact that the output has an alias annotation at all that affected the logic. ideally, having an alias annotation of a different shouldn't affect the parsing, that's not part of the grammar specification. e.g. Tensor(a) -> Tensor(b) should be the same as Tensor -> Tensor.

**Comment by bdhirsh** [comment]
>  e.g. Tensor(a) -> Tensor(b) should be the same as Tensor -> Tensor.

Hmm maybe I'm not following. Doesn't that imply that:

`Tensor(a) -> Tensor` should be the same as `Tensor -> Tensor` too?

Then for this schema:
`(Tensor(a -> *) -> Tensor[]`

Is the `a -> *` annotation for the argument meant to imply that the return aliases the argument? That's not super intuitive to me, but I also don't know as much about how torchscript deals with aliasing.

**Review by Krovatkin — APPROVED** [review]
LGTM

**Comment by eellison** [comment]
> Tensor(a) -> Tensor should be the same as Tensor -> Tensor too?

Yes it does. Per aliasing specification.  

> (Tensor(a -> *) -> Tensor[]

I agree that it's not super intuitive, but see comment above:

...

Yea, I am not saying this is the ideal syntax but updating it would require:
- parsing and storing contained alias sets, which we would have to change in two different parsers because the original aliasing parser is now duplicated in python (thus the changes i am making here)
and then either
- tracking the contained elements of containers and not automatically be the wildcard set. that would be a huge change and involve a number of schema, aliasdb, and memory dag changes
- logic to see if an annotated alias set has now been used in a contained alias set which would imply that it has entered the wildcard set.
I think the 2 would be a good medium term change but together is a pretty good changeset for what is a bug and effects memory management etc in SR, autodiff correctness for JIT, etc, and this PR fixes the bug in a way that makes the schemas compatible with the other existing schemas today.

File: tools/codegen/model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eellison) [reply tools/codegen/model.py:924]: Yea, in this PR
> We assume lists contain memory which aliases the heap

I think there are a number of follow up issues to unify and/improve the aliasing specification and particularly the conformance of both autograd's understanding of it and JIT's. I knew that there was a re-implementation of aliasing parsing happening in python but was not really aware of the details...

**Comment by bdhirsh** [comment]
@eellison let me know if you disagree, but to me `Tensor(a) -> Tensor(a)[]` seems like the "correct" frontend alias annotations for an op like `torch.split`. The downside of changing the annotations in the frontend in this way is that it's a breaking change for any code that relies on the existing aliasing semantics, including external code:
- the codegen (manually fixed here)
- the autograd not-implemented fallback (manually fixed here)
- any external hardware vendors that rely on the existing aliasing semantics externally (I think for example, MLC might rely on this cc @albanD)

These all would no longer be able to trust the alias annotations of the return arguments, which doesn't feel like the correct end-state.

If the existing alias annotations result in a bug for torchscript/SR, then that sound like something we should try to work around directly in the torchscript alias handling?

> yea, I am not saying this is the ideal syntax but updating it would require...

You pointed out that "really fixing it" would require a pretty huge change. If we agree that the existing bug is in torchscript's aliasing machinery, then I wonder if we can add a simpler bandaid by maybe transforming the aliasing annotations under the hood for torchscript? E.g. literally converting `Tensor(a) -> Tensor(a)[]` to `Tensor(a -> *) -> Tensor[]` at the time that torchscript parses aliasing annotations.

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

**Comment by eellison** [comment]
> Tensor(a) -> Tensor(a)[] seems like the "correct" frontend

Yea, i agree with that, but this annotation is being ignored both in every parser that exists right now. Something like `Tensor(a) -> Tensor(b)[]` would result in the same exact behavior with respect to the current handling. The `Tensor(a)[]` is being completely ignored, and instead the only thing that is being read is the presence of an annotation at all. 

If we want to improve the aliasing grammar's ability to specify things, I'm all for it, but this annotation syntax was added without the corresponding changes to the grammar that would actually enable the ideal state that you are talking about.

**Comment by albanD** [comment]
> Yea, i agree with that, but this annotation is being ignored both in every parser that exists right now (at least within core). Something like Tensor(a) -> Tensor(b)[] would result in the same exact behavior with respect to the current handling. The Tensor(a)[] is being completely ignored, and instead the only thing that is being read is the presence of an annotation at all.

Do you mean that all the existing fallback are buggy and not doing the right thing in the context of Tensor[] ?
A quick test of the conjugate view one seems to show that it works just fine right now:
```python
import torch

a = torch.rand(3, 3, dtype=torch.complex64)
print(a)

b = a.conj()

c = b.unbind()
print(c)
```

**Comment by eellison** [comment]
@albanD the aliasing specification within native_functions.yaml is “if there is an annotation, assume things alias”. Which seems to work ! But is not the same as implementing parsing of contained alias sets as described by Brian above. 

Also, see operators like https://github.com/pytorch/pytorch/issues/30526

**Comment by bdhirsh** [comment]
From talking offline with @albanD, I think we agree that the main constraint is that we don't want to BC-break any boxed fallback code, including code from external backends. I think this PR as is actually breaks the pytorch/xla CPU fallback (although it doesn't show up in CI) [here](https://github.com/pytorch/pytorch/blob/4d601a1c362a300450ecc3204eeca24cb17c0b38/aten/src/ATen/native/CPUFallback.cpp#L138). We also don't have a good way of checking / knowing what external code might broken. So I think the main contraint is that return arguments should continue show up as aliasing if they used to show up as aliasing.

> I can special case the aliasing behavior within alias_analysis.cpp if we want, fine, but that not a great state to be in IMO as it makes custom operators now divergent from native_functions.yaml ones.

Where does the divergent behavior come from for custom ops? My first thought would be that everything boils down to a `TORCH_LIBRARY` call (both custom and native ops), which goes through the same schema parsing machinery.

**Comment by eellison** [comment]
In this parser we do check that the alias_info are matching in this branch https://github.com/pytorch/pytorch/blob/4d601a1c362a300450ecc3204eeca24cb17c0b38/aten/src/ATen/native/CPUFallback.cpp#L151

but then do not check the same thing in the following branch https://github.com/pytorch/pytorch/blob/4d601a1c362a300450ecc3204eeca24cb17c0b38/aten/src/ATen/native/CPUFallback.cpp#L163 

I think that's because this is encoding the special meaning which is that outputs which have annotations are always assumed to alias, without parsing the contained `(a)`. I can retain that by adding the `Tensor(a)[]` output tensor annotation.

> Where does the divergent behavior come from for custom ops? My first thought would be that everything boils down to a TORCH_LIBRARY

If i were to add a custom aliasing logic in alias_analysis.cpp instead of making the change to the schemas as reflected here, that's where the divergence would come from.

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

File: aten/src/ATen/native/README.md [hunk 4]
```diff
@@ -250,6 +250,13 @@ There is also another situation in which we use annotations, namely views.
   - `transpose(Tensor(a) self, int dim0, int dim1) -> Tensor(a)`
     An alias to the memory represented by `self` may be also returned, however it is not mutated.
 
+When a Tensor views are contained in a Tensor list, we need to represent that the output list
+contains Tensors that alias the input.
+  - `func: chunk(Tensor(a -> *) self, int chunks, int dim=0) -> Tensor[]`
```
Comment (albanD) [inline aten/src/ATen/native/README.md:255]: nit the output schema needs updating?

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp [hunk 5]
```diff
@@ -64,36 +83,72 @@ void autogradNotImplementedFallbackImpl(const c10::OperatorHandle& op, c10::Disp
   is_inplace_output.reserve(num_returns);
   is_aliased_output.reserve(num_returns);
 
+  int aliased_input_idx = -1;
+  int aliased_output_idx = -1;
+
+  // if a Tensor is annotated with (a -> *) that implies that all lists of
```
Comment (albanD) [inline torch/csrc/autograd/autograd_not_implemented_fallback.cpp:89]: Can the changes to this file be reverted now?

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

File: test/cpp/api/autograd.cpp [hunk 6]
```diff
@@ -1220,4 +1220,4 @@ TEST(TestAutogradNotImplementedFallback, TensorlistOp) {
 // test_custom_autograd_repeated_grad_grad
 // test_return_leaf
 // test_anomaly_detect_nan
-// test_no_grad_copy
+// test_no_grad_copy
```
Comment (albanD) [inline test/cpp/api/autograd.cpp:1223]: nit missing newline?

**Review by albanD — APPROVED** [review]
LGTM
Some newlines at the end of files to make lint happy only.

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

**Pull request closed without merging**